### PR TITLE
feat: verify_proof

### PR DESCRIPTION
**Pull request by LennyTaelman**
Stacked on #75. Implements `verify_proof` in utils/verify.py which takes a repository, sorry location, proof string and verifies that the proof compiles. This is a bit hacked using REPL, it does not do any direct meta-programming or type checking. 

Could use some stress-testing, in particular needs to be tested for false positives against an example where a REPL tactic returns empty list of goals. 

Also includes a test in `tests/test_verify.py`, using a mock repo, list of proofs, list of false proofs.

### CODE REVIEW

**Comment by LennyTaelman** [comment]
Summary of how this works:
- make modified version of file excising the "sorry" string and replacing it with the proof string
- run both original and modified file through REPL, collecting sorries with their coordinates and goal strings
- if REPL reports any errors, fail (this is done in repl.read_file)
- verify that modified file has exactly 1 fewer sorries
- compute expected location of the remaining sorries, and verify that they exist in the modified file and have the same goal

File: sorrydb/utils/verify.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    Verify if a proof successfully replaces a sorry at a specific location.
+
+    Args:
+        repo_dir: Path to the repository
+        lean_version: Lean version tag
+        location: Dictionary containing sorry location info (file and coordinates)
+        proof: The proof string to replace the sorry
+
+    Returns:
+        Boolean indicating whether the proof successfully replaces the sorry
```
Comment (austinletson) [inline sorrydb/utils/verify.py:24]: There are a lot of reasons why `verify_proof` would return `False`. 

It seems useful in general for the caller to have more information about why `verify_proof` was unable to verify the proof. The difference between being unable to run Repl is quiet different from the goals not agreeing.

One option would be to create a `VerificationError` exception and then some subclasses for more specfic exceptions, e.g., `NoSorryRemove` and `MismatchedGoals`. These would allow the caller to treat different cases differently.

File: sorrydb/utils/verify.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    start_index = position_to_index(
+        original_file, location["start_line"], location["start_column"]
+    )
+    end_index = position_to_index(
+        original_file, location["end_line"], location["end_column"]
+    )
+
+    # Get the actual sorry text to verify we're replacing the right thing
+    sorry_text = original_file[start_index:end_index]
+    logger.info(f"Sorry text to be excised: {sorry_text}")
```
[the remark this answers is not in the document]
Comment (austinletson) [reply sorrydb/utils/verify.py:52]: Ah, there are potentially different sorry indicators. Maybe we should hold a list of these somewhere to use in cases like this?

File: pyproject.toml [hunk 3]
```diff
@@ -5,7 +5,7 @@ description = "Build a continously updating database of sorry statments in publi
 authors = [{ name = "Your Name", email = "you@example.com" }]
 license = { text = "Apache License" }
 readme = "README.md"
-requires-python = ">=3.13"
+requires-python = ">=3.12"
```
[the remark this answers is not in the document]
Comment (LennyTaelman) [reply pyproject.toml:8]: But I guess it is safe to leave it in for now. (And agreed: since we want to make it as easy as possible for others to use and adapt the code, it is best not to have too stringent requirements).

File: sorrydb/utils/verify.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (LennyTaelman) [reply sorrydb/utils/verify.py:52]: This can also be removed. It was meant for debugging (e.g. at first I was removing the shifted text " orry"). In any case, one of our tests will now easily catch this.

File: sorrydb/utils/verify.py (hunk 1 above)
Comment (LennyTaelman) [reply sorrydb/utils/verify.py:24]: Good point, I'll make an issue about error handling in general. Since we call a lot of external tools, and there are many layers where things could go wrong (but we do want to continue processing other sorries/branches/repositories), it may be good to think a bit about a uniform strategy and naming convention for exceptions...

File: sorrydb/utils/verify.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+import tempfile
+from pathlib import Path
+from typing import Dict, List, Optional
+
+from .repl_ops import LeanRepl, setup_repl
+
+logger = logging.getLogger(__name__)
+
+
+def verify_proof(repo_dir: Path, lean_version: str, location: Dict, proof: str) -> bool:
```
[the remark this answers is not in the document]
Comment (LennyTaelman) [reply sorrydb/utils/verify.py:13]: Yes, this makes perfect sense.

File: sorrydb/utils/verify.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    ) as tmp:
+        tmp.write(modified_file.encode("utf-8"))
+        tmp.flush()  # Ensure all data is written to disk
+
+        # Get the relative path from repo_dir to the temp file
+        temp_path = Path(tmp.name)
+        modified_file_path = temp_path.relative_to(repo_dir)
+
+        # Offset for sorries after the replaced one
+        offset = len(proof) - len(sorry_text)
```
[the remark this answers is not in the document]
Comment (LennyTaelman) [reply sorrydb/utils/verify.py]: Agreed.

File: tests/test_gather_sorries.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                        },
+                        "goal": sorry["goal"],
+                        "proof_state_id": sorry["proof_state_id"],
+                    }
+                    all_sorries.append(sorry_entry)
+
+    return all_sorries
+
+
+def test_sorries_match():
```
Comment (austinletson) [inline tests/test_gather_sorries.py:70]: What code is this test for? It seems like it is testing the `gather_sorries_from_repo` function defined in this test file.

File: tests/test_gather_sorries.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        toolchain_content = toolchain_path.read_text().strip()
+        if ":" in toolchain_content:
+            return toolchain_content.split(":", 1)[1]
+        return None
+    except Exception as e:
+        logger.error(f"Error reading lean-toolchain file: {e}")
+        return None
+
+
+def gather_sorries_from_repo(repo_dir: Path) -> list:
```
Comment (austinletson) [inline tests/test_gather_sorries.py:31]: This function seems like a reimplementation of `process_sorries.process_lean_repo`. Could we use that function?

File: tests/test_gather_sorries.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+import pytest
+from utils.repl_ops import LeanRepl, setup_repl
+
+# Configure logging
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def get_lean_version(repo_dir: Path) -> str:
```
Comment (austinletson) [inline tests/test_gather_sorries.py:16]: This seems like a reimplementation of `process_sorries.get_repo_lean_version`. Can we use that function?

File: tests/test_repl_read_file.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/env python3
```
Comment (austinletson) [inline tests/test_repl_read_file.py:1]: This file is a duplicate of `test_gather_sorries.py`

File: tests/test_verify.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env python3
+
+import json
+import logging
+from pathlib import Path
+
+import pytest
+from utils.verify import verify_proof
```
Comment (austinletson) [inline tests/test_verify.py:8]: These import needs to be prefixed with `sorrydb`: `from sorrydb.utils.verify ...`

File: tests/test_verify.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    - all proofs in mock_lean_repository_proofs.json are valid
+    - all proofs in mock_lean_repository_non_proofs.json are invalid
+    """
+
+    # Get the mock repository directory and proofs file
+    repo_dir = Path(__file__).parent / "mock_lean_repository"
+    proofs_file = Path(__file__).parent / "mock_lean_repository_proofs.json"
+    non_proofs_file = Path(__file__).parent / "mock_lean_repository_non_proofs.json"
+
+    # Get Lean version from lean-toolchain
```
Comment (austinletson) [inline tests/test_verify.py:22]: I think we should either hard code the lean version for the test or use the existing `process_sorries.get_repo_lean_version`

File: tests/test_verify.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+def test_verify_proofs():
+    """Test that
+    - all proofs in mock_lean_repository_proofs.json are valid
+    - all proofs in mock_lean_repository_non_proofs.json are invalid
+    """
+
+    # Get the mock repository directory and proofs file
+    repo_dir = Path(__file__).parent / "mock_lean_repository"
+    proofs_file = Path(__file__).parent / "mock_lean_repository_proofs.json"
```
Comment (austinletson) [inline tests/test_verify.py:19]: Could we move both of these json files into the `mock_lean_repository` directory?

I think we are going to eventually create more than one mock lean repository for testing and this will help keep things organized.

**Review by austinletson — CHANGES_REQUESTED** [review]
I played with the `test_verify_proofs` test some by editing the proofs and non proofs jsons and it worked as expected.

The `test_gather_sorries` and `test_repl_read_file` I don't fully understand. Could you provide some more information on what these are trying to test?

File: tests/test_verify.py (hunk 10 above)
Comment (LennyTaelman) [reply tests/test_verify.py]: Thanks!

File: tests/test_verify.py (hunk 11 above)
Comment (LennyTaelman) [reply tests/test_verify.py:20]: Agreed! Now uses the get_repo_lean_version

File: tests/test_verify.py (hunk 12 above)
Comment (LennyTaelman) [reply tests/test_verify.py]: Good point. I've moved them into the repo, and given them generic names (proofs.json, non_proofs.json) which can also be used by other mock repositories.

File: tests/test_gather_sorries.py (hunk 7 above)
Comment (LennyTaelman) [reply tests/test_gather_sorries.py:31]: We probably should, but that will require some more work. Right now it doesn't work since the mock repo is not a git repo, and hence git_blame fails on it.

Comment (LennyTaelman) [reply tests/test_gather_sorries.py:31]: I don't see a really clean solution now. I'll remove the test from this PR, but keep the "sorries.json" file in the mock repository for now.

**Comment by LennyTaelman** [comment]
Thanks for pointing out the mess with the gather_sorries, read_file test functions. These were far from ready, and I've now removed them from the PR.

There is a `sorries.json` file in the mock repository, and at some point we could use that for a test of e.g. process_lean_repo. But note that the mock repo is not a git repo, so things like `blame date` won't work.

File: sorrydb/utils/verify.py (hunk 1 above)
Comment (LennyTaelman) [reply sorrydb/utils/verify.py:24]: This is now handled in an upcoming PR which is stacked on this one.

**Comment by austinletson** [comment]
> Thanks for pointing out the mess with the gather_sorries, read_file test functions. These were far from ready, and I've now removed them from the PR.
> 
> There is a `sorries.json` file in the mock repository, and at some point we could use that for a test of e.g. process_lean_repo. But note that the mock repo is not a git repo, so things like `blame date` won't work.

Yes, currently we have some test coverage of `process_lean_repo` through the `update_database` tests, however these currently use the mock repos hosted separately on GitHub. Long term we should move as much testing as possible to mock repos held in the SorryDB repo like you have done here. This is a great foundation to improve those other tests!

**Comment by austinletson** [comment]
LGTM! @LennyTaelman I will let you to merge this.

**Review by austinletson — APPROVED** [review]
